src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs

```diff
@@ -2514,18 +2536,144 @@ public void Dispose()
                 return;
             }
 
-            this.DisposingEvent.SafeInvoke(this, EventArgs.Empty);
-            commandRuntime = null;
-            currentObjectInPipeline = null;
-            _input.Clear();
-            // _scriptBlock = null;
-            // _localsTuple = null;
-            // _functionContext = null;
+            try
+            {
+                if (_scriptBlock.HasCleanupBlock)
+                {
+                    InvokeCleanupBlock();
+                }
+            }
+            finally
+            {
+                this.DisposingEvent.SafeInvoke(this, EventArgs.Empty);
+
+                commandRuntime = null;
+                currentObjectInPipeline = null;
+                _input.Clear();
 
-            base.InternalDispose(true);
-            _disposed = true;
+                base.InternalDispose(true);
+                _disposed = true;
+            }
+        }
+
+        private void InvokeCleanupBlock()
+        {
+            ExecutionContext currentContext = LocalPipeline.GetExecutionContextFromTLS();
+            if (Context != currentContext)
+            {
+                // Something went wrong; Cleanup {} is being called from the wrong thread.
+                // Create an event to call it from the correct thread.
+                InvokeCleanupInOriginalContext();
+                return;
+            }
+
+            Pipe oldOutputPipe = _functionContext?._outputPipe;
+            try
+            {
+                Action<FunctionContext> cleanupBlock = _runOptimized
+                    ? _scriptBlock.CleanupBlock
+                    : _scriptBlock.UnoptimizedCleanupBlock;
+
+                if (_functionContext != null)
+                {
+                    _functionContext._outputPipe = new Pipe
+                    {
+                        NullPipe = true
+                    };
+                }
+
+                // run with no pipeline input or $input enumerator
+                RunClause(cleanupBlock, AutomationNull.Value, AutomationNull.Value);
+            }
+            finally
+            {
+                if (_functionContext != null)
+                {
+                    _functionContext._outputPipe = oldOutputPipe;
+                }
+            }
         }
 
         #endregion IDispose
+
+        #region Eventing for Cleanup
+
+        private void InvokeCleanupInOriginalContext()
```

Comment:
> If we don't depend on `Dispose` for invoking the `Cleanup` block,

We just need to add one more step in pipeline after End(). It will even simplify everything.

Reply:
It's not _quite_ that simple; `cleanup` is supposed to invoke regardless of exceptions in the other blocks. I think I know where it can be invoked to make it work that way, but it might be better to wait until Dongbo's had a chance to properly look through the pipeline processor code; he might have some specific recommendations there that could change where/how it needs to be invoked.

Reply:
> It's not quite that simple; cleanup is supposed to invoke regardless of exceptions in the other blocks. 

I see. I'd try to wrap the blocks in try-catch on as low level as possible because we call RestorePreviousScope() in finally clause on more high level.

Reply:
Aye, that's one of the issues I had a heck of a lot of trouble with initially, for sure. Keeping it all in the same scope is tricky at best.